module/epochs/epoch_lookup_test.go

```diff
@@ -1,65 +1,255 @@
+func (suite *EpochLookupSuite) CreateAndStartEpochLookup() {
+	lookup, err := NewEpochLookup(suite.state)
+	suite.Require().NoError(err)
+	ctx, cancel, errCh := irrecoverable.WithSignallerAndCancel(context.Background())
+	lookup.Start(ctx)
+	go unittest.FailOnIrrecoverableError(suite.T(), ctx.Done(), errCh)
+
+	suite.lookup = lookup
+	suite.cancel = cancel
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a current epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_Curr() {
+	epochs := []epochRange{suite.currEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a previous and current epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_PrevCurr() {
+	epochs := []epochRange{suite.prevEpoch, suite.currEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
 
-	for _, tc := range testCases {
-		epoch, err := lookup.EpochForView(tc.view)
-		require.NoError(t, err)
-		require.Equal(t, tc.epoch, epoch)
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a current and next epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_CurrNext() {
+	epochs := []epochRange{suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a previous, current, and next epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_CurrNextPrev() {
+	epochs := []epochRange{suite.prevEpoch, suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of epoch fallback triggered.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_EpochFallbackTriggered() {
+	epochs := []epochRange{suite.prevEpoch, suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+func (suite *EpochLookupSuite) TestProtocolEvents_EpochFallbackTriggered() {
+	// initially, only current epoch is committed
+	suite.CommitEpochs(suite.currEpoch)
+	suite.CreateAndStartEpochLookup()
+
+	// trigger epoch fallback
+	suite.epochFallbackTriggered = true
```

Comment:
## 🤯 

I find it time consuming to convince myself that there is not data race here. `EpochLookup` concurrently processes Epoch events.
In this particular case, one requires detailed understanding of `EpochLookup` to determine that no data race is possible. In particular needs to figure out that `state.Params` is only accessed within the constructor, i.e. in the test's thread. I would argue that thereby the test relies too heavily on internal implementation details of `EpochLookup`.

Reply:
Good point - added concurrency protection for these fields in https://github.com/onflow/flow-go/pull/2927/commits/e22499eaecc465a98c996b3657798b6d0c09096c